projects/manage-supervision-and-delius/src/main/kotlin/uk/gov/justice/digital/hmpps/integrations/delius/user/entity/User.kt

```diff
@@ -218,20 +223,194 @@ data class Caseload(
+    """
+)
+@Synchronize("contact")
+data class NextAppointment(
+
+    @Id
+    @Column(name = "contact_id")
+    val id: Long,
+
+    @Column(name = "offender_id")
+    val personId: Long,
+
+    @Column(name = "staff_id")
+    val staffId: Long,
+
+    @ManyToOne
+    @JoinColumn(name = "contact_type_id")
+    val type: ContactType,
+
+    @Column(name = "appointment_datetime")
+    val appointmentDatetime: ZonedDateTime,
+
+    )
+
+@Entity
+@Subselect(
+    """
+        select * from (
+            select 
+            q.staff_id,
+            q.offender_id,
+            q.appointment_datetime,
+            q.contact_id,
+            q.contact_type_id,
+            row_number() over (partition by staff_id, offender_id order by appointment_datetime desc) as row_num 
+            from
+            (
+                select 
+                to_timestamp(to_char(c.contact_date, 'yyyy-mm-dd') || ' ' || to_char(c.contact_start_time, 'hh24:mi:ss'), 'yyyy-mm-dd hh24:mi:ss') as appointment_datetime,
+                c.contact_id, 
+                c.contact_type_id,
+                c.staff_id,
+                c.offender_id
+                from contact c
+                join r_contact_type ct on c.contact_type_id = ct.contact_type_id and ct.attendance_contact = 'Y'
+                where c.CONTACT_START_TIME is not null
+                and c.soft_deleted = 0
+            ) q
+            where appointment_datetime < current_timestamp
+        )
+        where row_num = 1
+    """
+)
+@Synchronize("contact")
+data class PreviousAppointment(
+
+    @Id
+    @Column(name = "contact_id")
+    val id: Long,
+
+    @Column(name = "offender_id")
+    val personId: Long,
+
+    @Column(name = "staff_id")
+    val staffId: Long,
+
+    @ManyToOne
+    @JoinColumn(name = "contact_type_id")
+    val type: ContactType,
+
+    @Column(name = "appointment_datetime")
+    val appointmentDatetime: ZonedDateTime,
+
+    )
+
 interface CaseloadRepository : JpaRepository<Caseload, Long> {
     @Query(
         """
         select c from Caseload c
         join fetch c.person p
         join fetch c.team t
+        left join fetch c.nextAppointment na
+        left join fetch c.previousAppointment pa
         where c.staff.code = :staffCode
     """
     )
     fun findByStaffCode(staffCode: String, pageable: Pageable): Page<Caseload>
 
+    @Query(
+        """
+        select c from Caseload c
+        join fetch c.person p
+        join fetch c.team t
+        left join fetch c.nextAppointment na
+        left join fetch na.type naType
+        left join fetch c.previousAppointment pa
+        left join fetch pa.type paType
+        left join fetch c.latestSentence ls
+        left join fetch ls.mainOffence mo
+        left join fetch mo.offence moo
+        where c.staff.code = :staffCode
+        and ((:nameOrCrn is null or upper(p.crn) like '%' || upper(:nameOrCrn) || '%' or upper(p.forename || ' ' || p.surname) like '%' || upper(:nameOrCrn) || '%'))
```

Comment:
Should this include a filter on surname+forename too?  E.g. if someone searched for "Smith, John"?
```suggestion
        and (
          :nameOrCrn is null 
          or upper(p.crn) like '%' || upper(:nameOrCrn) || '%' 
          or upper(p.forename || ' ' || p.surname) like '%' || upper(:nameOrCrn) || '%'
          or upper(p.surname || ' ' || p.forename) like '%' || upper(:nameOrCrn) || '%'
          or upper(p.surname || ', ' || p.forename) like '%' || upper(:nameOrCrn) || '%')
```

Feel free to ignore this if you think it makes the query too verbose though.

Reply:
agreed! i like it. Thanks. Done